qt/widgets/instrumentview/src/InstrumentRenderer.cpp

```diff
@@ -0,0 +1,439 @@
+    auto type = componentInfo.componentType(i);
+    if (type == Mantid::Beamline::ComponentType::Rectangular ||
+        type == Mantid::Beamline::ComponentType::OutlineComposite) {
+      m_textureIndices.push_back(i);
+      m_reverseTextureIndexMap[i] = numTextures;
+      numTextures++;
+    }
+  }
+
+  if (numTextures > 0) {
+    m_textureIDs.resize(numTextures, 0);
+    colorTextures.resize(numTextures);
+    pickTextures.resize(numTextures);
+  }
+}
+
+InstrumentRenderer::~InstrumentRenderer() {
+  for (size_t i = 0; i < 2; ++i) {
+    if (m_displayListId[i] != 0) {
+      glDeleteLists(m_displayListId[i], 1);
+    }
+  }
+}
+
+void InstrumentRenderer::renderInstrument(const std::vector<bool> &visibleComps,
+                                          bool showGuides, bool picking) {
+  if (std::none_of(visibleComps.cbegin(), visibleComps.cend(),
+                   [](bool visible) { return visible; }))
+    return;
+
+  OpenGLError::check("InstrumentActor::draw()");
+  size_t i = picking ? 1 : 0;
+  if (m_useDisplayList[i]) {
+    glCallList(m_displayListId[i]);
+  } else {
+    m_displayListId[i] = glGenLists(1);
+    m_useDisplayList[i] = true;
+    glNewList(m_displayListId[i],
+              GL_COMPILE); // Construct display list for object representation
+    draw(visibleComps, showGuides, picking);
+    glEndList();
+    if (glGetError() == GL_OUT_OF_MEMORY) // Throw an exception
+      throw Mantid::Kernel::Exception::OpenGLError(
+          "OpenGL: Out of video memory");
+    glCallList(m_displayListId[i]);
+  }
+  OpenGLError::check("InstrumentActor::draw()");
+}
+
+void InstrumentRenderer::draw(const std::vector<bool> &visibleComps,
+                              bool showGuides, bool picking) {
+  const auto &compInfo = m_actor.componentInfo();
+  std::vector<bool> visited(compInfo.size(), false);
+
+  for (size_t i = compInfo.root(); i != std::numeric_limits<size_t>::max();
+       --i) {
+    auto type = compInfo.componentType(i);
+    if (type == Mantid::Beamline::ComponentType::Infinite)
+      continue;
+
+    if (type == Mantid::Beamline::ComponentType::Rectangular) {
+      visited[i] = true;
+      updateVisited(visited, compInfo.children(i));
+      if (visibleComps[i])
+        drawRectangularBank(i, picking);
+      continue;
+    }
+
+    if (type == Mantid::Beamline::ComponentType::OutlineComposite) {
+      visited[i] = true;
+      updateVisited(visited, compInfo.children(i));
+      if (visibleComps[i])
+        drawTube(i, picking);
+      continue;
+    }
+
+    if (type == Mantid::Beamline::ComponentType::Structured) {
+      visited[i] = true;
+      updateVisited(visited, compInfo.children(i));
+      if (visibleComps[i])
+        drawStructuredBank(i, picking);
+      continue;
+    }
+
+    if (!compInfo.isDetector(i) && !showGuides) {
+      visited[i] = true;
+      continue;
+    }
+
+    if (compInfo.hasValidShape(i) && visibleComps[i] && !visited[i]) {
```

Comment:
FYI, you should be aware of how `ComponentInfo::boundingBox` deals with a very similar situation. Ranges of sub-detectors are tracked, not individual sub-detectors.

Reply:
Isn't this helped by the fact that component ranges are exposed at this level? (But not exposed to the outside world beyond `Geometry::ComponentInfo`

Reply:
Maybe `ComponentInfo` needs some iterators to help with this.

Reply:
Would performing this check improve the performance of this loop? It's `O(n)` for the number of components in the instrument. I guess it could be improved if there was a diminishing list? Or I keep track of the number of components traversed and then exit when that number matches the size of `ComponentInfo`?